### PR TITLE
KAFKA-9431: Expose API in KafkaStreams to fetch all local offset lags

### PR DESCRIPTION
**Pull request by vinothchandar**
- Adds KafkaStreams#allLocalOffsetLags(), which returns lag information of all active/standby tasks local to a streams instance
 - LagInfo class encapsulates the current position in the changelog, endoffset in the changelog and their difference as lag
 - Lag information is a mere estimate; it can over-estimate (source topic optimization), or under-estimate.
 - Each call to allLocalOffsetLags() generates a metadata call to Kafka brokers, so caution advised
 - Unit and Integration tests added.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by vinothchandar** [comment]
@vvcephei @brary Open this PR for lag related changes...

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 1]
```diff
@@ -1144,4 +1154,74 @@ public void cleanUp() {
...
+            allEndOffsets = adminClient.listOffsets(
+                allPartitions.stream()
+                    .collect(Collectors.toMap(Function.identity(), tp -> OffsetSpec.latest()))
+            ).all().get();
+        } catch (final RuntimeException | InterruptedException | ExecutionException e) {
+            throw new StreamsException("Unable to obtain end offsets from kafka", e);
+        }
+        log.debug("Current end offsets :{}", allEndOffsets);
+        for (final Map.Entry<TopicPartition, ListOffsetsResultInfo> entry : allEndOffsets.entrySet()) {
+            // Avoiding an extra admin API lookup by computing lags for not-yet-started restorations
```
Comment (vinothchandar) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:1274]: @vvcephei @brary I actually have a concern with this overestimation.. Let's take a typical streams deployment with num_standby_replicas =1 ... During the time period, where the lag is fetched when a standby changelog partition is assigned but has not yet processed any events, we will report a really high value that may not lead to IQs being failed (since the store is deemed laggy).. Given a caller that periodically fetches the positions every few seconds would have a non-zero past value already, wonder if it makes sense to not report this partition instead.. i.e we only report lags on store partitions which have begun consumption..  At least for ksqlDB, I am thinking servers can track easily when lag was reported for a given store partition, given server and decide on using that value based on how recently it was updated..  But the values we receive won't be 'jumpy'. End of the day, I am also theorizing how this will behave.. So this is more about which way we want to bias. 


Across the cluster, this time period will not in sync with each other and can happen differently. So not sure if this will in-fact provide correct relative order..

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java [hunk 2]
```diff
@@ -1144,4 +1154,74 @@ public void cleanUp() {
...
+     * Returns {@link LagInfo}, for all store partitions (active or standby) local to this Streams instance. Note that the
+     * values returned are just estimates and meant to be used for making soft decisions on whether the data in the store
+     * partition is fresh enough for querying.
+     *
+     * Note: Each invocation of this method issues a call to the Kafka brokers. Thus its advisable to limit the frequency
+     * of invocation to once every few seconds.
+     *
+     * @return map of store names to another map of partition to {@link LagInfo}s
+     */
+    public Map<String, Map<Integer, LagInfo>> allLocalStorePartitionLags() {
```
Comment (vinothchandar) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:1230]: is there a notion of `experimental` APIs in streams? (like some other apache projects like Spark have) I think we can mark this experimental if so, setting the expectations with the user for the next release.

Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:1230]: There's not. There's "interface: evolving", but I think it's only for... interfaces. We already say in the javadoc that the returned values are estimates, so I think we've created enough wiggle room in the correctness of the returned lags. As to whether the method signature itself might change in the future, we'll just have to go the normal route of deprecating and replacing if we want to change it.

**Comment by vinothchandar** [comment]
First system tests run: passes.

http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2020-01-15--001.1579123047--vinothchandar--kafka-9431-local-lag-fetch-api--5ab2bee/report.html

File: clients/src/test/java/org/apache/kafka/clients/admin/MockAdminClient.java [hunk 3]
```diff
@@ -450,12 +450,17 @@ public ListPartitionReassignmentsResult listPartitionReassignments(Optional<Set<
...
 
     @Override
     public ListOffsetsResult listOffsets(Map<TopicPartition, OffsetSpec> topicPartitionOffsets, ListOffsetsOptions options) {
-        throw new UnsupportedOperationException("Not implement yet");
+        throw new UnsupportedOperationException("Not implemented yet");
+    }
+
+    @Override
+    public ListOffsetsResult listOffsets(Map<TopicPartition, OffsetSpec> topicPartitionOffsets) {
+        throw new UnsupportedOperationException("Not implemented yet");
```
Comment (vvcephei) [inline clients/src/test/java/org/apache/kafka/clients/admin/MockAdminClient.java]: What is the purpose of this change? It _looks_ unrelated, but I'm guessing it's related somehow to your tests?

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 1 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:1274]: Thanks for bringing this up. I'm not totally sure I follow, though...

Let's say the partition in question has an end offset of 100 and a true "earliest" offset of 90 (due to compaction). If an instance has this partition assigned but has not yet recorded any position, we'd report its lag as 100. Another instance that has started processing but only processed one record would be reported with a lag of 9. This doesn't accurately represent the amount-of-work difference between the two instances (which is actually only one record, though the difference in lags is 91), but it _does_ correctly state that the second instance is ahead of the first in freshness.

I also seems reasonable to not report a lag at all for not-yet-started stores, but then you have to handle the special case in client-side code. I.e., the metadata API would tell you that this instance owns a store, but then there's no lag reported, so maybe there was a rebalance and we lost the store, or maybe we do own the store, but haven't started processing yet... It still seems simpler to me to say "woah, that lag is really big, like it's lagging by 100% of the topic, maybe I won't query this store" if I don't want to query really stale data, versus handling the special case to reach the same conclusion.

Realistically, even if we are still restoring, but the store is still behind by a year (for example), you _still_ wouldn't want to return the results, but rather would have _some_ heuristic of how stale is ok and how stale is too much, which would be able to handle this math just as easily either way.

One other thought is that offset values can still be jumpy, due to compaction. So there's nothing to say that there are any records at all between offset 5 and offset 9999. Maybe this is contributing to my feeling that it's ok to just gloss over this case.

File: streams/src/test/java/org/apache/kafka/streams/integration/LagFetchIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+@Category({IntegrationTest.class})
+public class LagFetchIntegrationTest {
+
+    @ClassRule
+    public static final EmbeddedKafkaCluster CLUSTER = new EmbeddedKafkaCluster(1);
+
+    private static final long CONSUMER_TIMEOUT_MS = 60000;
+
+    @Rule
+    public TemporaryFolder folder = new TemporaryFolder();
```
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/integration/LagFetchIntegrationTest.java]: This is neat, but we shouldn't use it. There's an IntegrationTestUtil for getting a temporary folder, which is hooked in to support for different testing environments to set their desired temporary file location.

**Review by vvcephei — COMMENTED** [review]
Thanks for all your work on this!

Looks good overall, just a few small comments.

One other thought I had was about swapping out those internal maps for a ConcurrentHashMap. Are any of these on a hot path, such that we would worry about impacting the performance of Streams if frequent IQs are contending with processing for the striped locks?

If it is a concern, I'm wondering if they are really necessary. It seems like their main purpose is as a memory barrier, but there are other ways to add memory barriers without the synchronization overhead.

WDYT?

**Comment by vvcephei** [comment]
Both builds failed, but it looks unrelated:
* https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/245/console
* https://builds.apache.org/job/kafka-pr-jdk11-scala2.13/4260/console

Retest this, please.

**Comment by vinothchandar** [comment]
Another run of System tests passed as well.
http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2020-01-15--001.1579149199--vinothchandar--kafka-9431-local-lag-fetch-api--cf7522e/report.html

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 2 above)
Comment (vinothchandar) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:1230]: Okay.. lets stick to what we have.

File: clients/src/test/java/org/apache/kafka/clients/admin/MockAdminClient.java (hunk 3 above)
Comment (vinothchandar) [reply clients/src/test/java/org/apache/kafka/clients/admin/MockAdminClient.java]: Thats the variant I am using in `allLocalOffsetLags`.. I swear, at some point I needed it for the KafkaStreamsTest to pass .. I took another look though and test currently with a partial mock on MockAdminClient seems to pass without this change.. So will back this out. Thanks for the call out

File: streams/src/test/java/org/apache/kafka/streams/integration/LagFetchIntegrationTest.java (hunk 4 above)
Comment (vinothchandar) [reply streams/src/test/java/org/apache/kafka/streams/integration/LagFetchIntegrationTest.java]: I see a `TestUtils#temporaryFolder` which basically uses java `Files.createTempDirectory/deleteOnExit` route.. This is what you switched to in the QueryableStateIntegrationTest as I see .. 
I ll play by house rules.. but I don't see a `IntegrationTestUtil` or a related method in `IntegrationTestUtils` .. I see one where it purges local state dir from streams config.. Is that what you are referring to.. 

Once you respond, I will clean up both tests

**Comment by vvcephei** [comment]
Hey @vinothchandar , I looked into the possible contention between IQ and processing, and I found that, yes, they may introduce contention, but it's probably not too much of a concern for this PR.

In most cases, these collections are only accessed during rebalance, which doesn't need to be super high performing. They are also accessed during restoration, which is more of a concern because we want the restore to complete as fast as possible. However, there would only be contention when we call the local lag API, so an easy workaround is just to call it less often if profiling shows high contention.

Also, this is a new API, so I'm not so concerned that existing code would suddenly perform worse. There will certainly be more synchronization overhead in the restore path, but there's also a lot of I/O in the restore path, and I'd generally expect the I/O to dominate.

Finally, I noticed that the collection most used in the hot processing path, `org.apache.kafka.streams.processor.internals.AssignedTasks#running`, was already a ConcurrentHashMap. 

In conclusion, it seems reasonable to say that we may be introducing more overhead here, but it's likely a small amount. In exchange for the new feature it's probably worthwhile.

Plus, there are concurrent efforts to completely refactor the internals, and those maps would probably go away in that case. It makes sense to go ahead and take a simple and safe approach with this change, see what happens with the refactor, and then optimize performance as necessary.

**Review by vinothchandar — COMMENTED** [review]
This method is documented to be called once every few seconds, not in a critical path.. There would be lot more problems on the Kafka brokers themselves, if someone calls this in a tight for-loop say. 
 
On the maps, the need for concurrent map is mainly to provide safe iteration of the map, while removals may be going on (my conclusion after reading code in places like AssignedTasks etc where the task state management seems to be removing and adding tasks from different maps).. If you have a better suggestion, I am listening.. 
Not sure I followed the memory barrier comment.. Ideally, if each task had just the two longs (position, endoffset) within themselves, we can just use `volatile` and deal with it.. Unfortunately, the tasks seem to have maps to deal with multiple partitions (partition grouper support?). I am intending to do a follow up, along with exploring fetching of the end offset out of memory... This is also some feedback for the re-design.. @guozhangwang  did confirm that the new design with have a clear lag api for Tasks.. 

On performance of concurrent map itself, I am not sure what exactly you are getting at. I have used it critical paths before with good success.. Can you elaborate your specific performance concern? (we can also sidebar this later, if you are happy to revisit later, given its only called every few seconds)

Note: I will respond to the larger open item on lag reporting once I am in office.

File: clients/src/test/java/org/apache/kafka/clients/admin/MockAdminClient.java (hunk 3 above)
Comment (vvcephei) [reply clients/src/test/java/org/apache/kafka/clients/admin/MockAdminClient.java]: Thanks!

File: streams/src/test/java/org/apache/kafka/streams/integration/LagFetchIntegrationTest.java (hunk 4 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/integration/LagFetchIntegrationTest.java]: Ah, yes, it's `org.apache.kafka.test.TestUtils#tempDirectory()`. My mistake. The protocol is for all temporary state in Kafka tests to use that method.

The change I made in `QueryableStateIntegrationTest` is basically what we should do here as well.

**Comment by vvcephei** [comment]
Hey @vinothchandar , Thanks for the response.

I think we're on the same page about the concurrency question.

Once you back out the MockAdminClient changes and fix the LagFetchIntegrationTest temp folder usage, we should be good to go!

Thanks again for all your work on this awesome feature.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java (hunk 1 above)
Comment (vinothchandar) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:1274]: >but it does correctly state that the second instance is ahead of the first in freshness.

Fair. Your argument is if you have not begun work, then you would be behind another instance that has begun work.. 

>reasonable to not report a lag at all for not-yet-started stores, but then you have to handle the special case in client-side code.

the client code is going to have to use some mechanism to "fence" really bad lag.. i.e full restorations.. My concern was that if we hit this race condition, you would suddenly see a jump to "oh this replica needs to fully restore" and then back to the actual lag..

>One other thought is that offset values can still be jumpy, due to compaction.

I did think about this a bit.. I think what we have here would work fairly well in the common case, where the active and standby are apart by lag within the `uncompacted` portion of the changelog topic.. 

> So there's nothing to say that there are any records at all between offset 5 and offset 9999.
Once beyond that, since we just rely on the latest lag value, this is okay.. if there are no records, the restoration will be faster than expected.. but as a relative metric, it's still good.. 

In all, I am okay sticking with this behavior for now.. We can revisit based on real world experience.. and really build a solid solid implementation in the next month or so..

**Comment by vinothchandar** [comment]
@vvcephei CI passed!

**Review by vvcephei — APPROVED** [review]

File: streams/src/main/java/org/apache/kafka/streams/LagInfo.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    @Override
+    public boolean equals(final Object obj) {
+        if (!(obj instanceof LagInfo)) {
+            return false;
+        }
+        final LagInfo other = (LagInfo) obj;
+        return currentOffsetPosition == other.currentOffsetPosition
+            && endOffsetPosition == other.endOffsetPosition
+            && this.offsetLag == other.offsetLag;
```
Comment (tedyu) [inline streams/src/main/java/org/apache/kafka/streams/LagInfo.java:75]: Since offsetLag is computed from the other two fields, this comparison is not necessary.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java [hunk 6]
```diff
@@ -46,7 +47,7 @@
     private final StateRestoreListener userStateRestoreListener;
     private final Map<TopicPartition, Long> restoreToOffsets = new HashMap<>();
     private final Map<String, List<PartitionInfo>> partitionInfo = new HashMap<>();
-    private final Map<TopicPartition, StateRestorer> stateRestorers = new HashMap<>();
+    private final Map<TopicPartition, StateRestorer> stateRestorers = new ConcurrentHashMap<>();
```
Comment (tedyu) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:50]: I noticed some maps are changed to ConcurrentHashMap.
May I ask what was the selection criterion for the change ?

thanks

Comment (vinothchandar) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:50]: This is for safe iteration from the thread calling the lag fetch API
